### PR TITLE
[Fix #3779] Add Rails/FilePath cop

### LINKED ISSUE #3779: Rails Cop Idea: `Rails.root.join`
Would love a cop that specifies to always use syntax:

```ruby
Rails.root.join('app', 'models', 'goober') # good
```

over:

```ruby
# bad
Rails.root.join('app/models/goober')
"#{Rails.root}/app/models/goober"
File.join(Rails.root, 'app/models/goober')
# etc...
```

We've got a bunch of formats littered throughout our codebase.

### PR DESCRIPTION
**Pull request by iguchi1124**
**Add Rails/FilePath cop**

Related issue: #3779

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: spec/rubocop/cop/rails/file_path_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Rails::FilePath do
+  subject(:cop) { described_class.new }
+
+  context 'when Rails.root.join with some path strings' do
+    let(:source) { 'Rails.root.join(\'app\', \'models\', \'user.rb\')' }
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/file_path_spec.rb]: I'd use "" to eliminate the need to escape ' inside the string. We should actually have a cop for this! :-) /cc @Drenmi @pocke

File: spec/rubocop/cop/rails/file_path_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    it 'registers an offense' do
+      inspect_source(cop, source)
+      expect(cop.offenses.size).to eq(0)
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/file_path_spec.rb]: Prefer the usage of `be_empty`.

File: lib/rubocop/cop/rails/file_path.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # @example
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/file_path.rb]: I'd also add some general explanations here.

File: spec/rubocop/cop/rails/file_path_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Rails::FilePath do
+  subject(:cop) { described_class.new }
+
+  context 'when Rails.root.join with some path strings' do
+    let(:source) { 'Rails.root.join(\'app\', \'models\', \'user.rb\')' }
+
+    it 'registers an offense' do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/file_path_spec.rb]: does not register :-)

File: spec/rubocop/cop/rails/file_path_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Rails::FilePath do
+  subject(:cop) { described_class.new }
+
+  context 'when Rails.root.join with some path strings' do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/file_path_spec.rb]: when using

File: spec/rubocop/cop/rails/file_path_spec.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    let(:source) { 'Rails.root.join(\'app\', \'models\', \'user.rb\')' }
+
+    it 'registers an offense' do
+      inspect_source(cop, source)
+      expect(cop.offenses.size).to eq(0)
+    end
+  end
+
+  context 'when File.join with Rails.root' do
```
Comment (bbatsov) [inline spec/rubocop/cop/rails/file_path_spec.rb]: when using

File: lib/rubocop/cop/rails/file_path.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          return unless rails_root_join_nodes(node).map(&:method_args)
+                                                   .flatten
+                                                   .any? do |arg|
+                                                     arg.source =~ %r{/}
+                                                   end
+
+          offense(node)
+        end
+
+        def offense(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/file_path.rb]: I'd name this `register_offense` or something like this.

File: lib/rubocop/cop/rails/file_path.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        end
+
+        def on_send(node)
+          file_join_with_rails_root(node)
+          rails_root_join_with_slash_separated_path(node)
+        end
+
+        private
+
+        def file_join_with_rails_root(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/file_path.rb]: I'd name this `check_for_file_join_with_rails_root` or something else. Or maybe you could extract a separate predicate method from this one to simplify the code?

File: lib/rubocop/cop/rails/file_path.rb [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+          return unless file_join_nodes?(node)
+          return unless file_join_nodes(node).map(&:method_args)
+                                             .flatten
+                                             .any? { |e| rails_root_nodes?(e) }
+
+          offense(node)
+        end
+
+        def rails_root_join_with_slash_separated_path(node)
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/file_path.rb]: Same here.

**Comment by bbatsov** [comment]
You should also write a matching rule in the Rails style guide.

**Comment by iguchi1124** [comment]
@bbatsov

> You should also write a matching rule in the Rails style guide.

OK, I will.
where kind of table content should I write in to? by the way.

https://github.com/bbatsov/rails-style-guide#table-of-contents

File: spec/rubocop/cop/rails/file_path_spec.rb (hunk 1 above)
Comment (Drenmi) [inline spec/rubocop/cop/rails/file_path_spec.rb]: I had this exact thing in a code review last week, so I guess I agree that `Style/StringLiterals` should account for this. 😅

Comment (iguchi1124) [inline spec/rubocop/cop/rails/file_path_spec.rb]: I'll create this issue!

**Comment by iguchi1124** [comment]
Thank you for great reviews and i fixed all of review comment point.

**Comment by bbatsov** [comment]
> where kind of table content should I write in to?

Guess such code most often comes up in configuration, so I guess it can be in this section...

**Comment by dominicsayers** [comment]
@Bodacious You're hard coding the path separator. It's reasonable assumption that it will be a "/" but not 100% universal. One somebody might want to run your code on Windows or a future OS that doesn't use "/".

Regarding the ugliness of the first format, I sometimes use

```
Rails.root.join(*%w[one two three four five.rb])
```

which, while still a bit ugly, is at least shorter.

**Comment by Drenmi** [comment]
@dominicsayers: Care to open a new request to add a configuration option for the path separator? 🙂

**Comment by dominicsayers** [comment]
You mean remove the [hard-coded](https://github.com/bbatsov/rubocop/blob/v0.51.0/lib/rubocop/cop/rails/file_path.rb#L60) forward slash from the cop? That wouldn't need a configuration option, just replace the slash with `File::SEPARATOR`

**Comment by Bodacious** [comment]
@dominicsayers thanks for taking the time to explain it. I never considered OS that don't use "/" 👍

**Comment by Drenmi** [comment]
> That wouldn't need a configuration option [...]

Ah. Yes. Of course. I think we have a Windows CI now, so I might give it a shot.

**Pull request merged**